dataframe.py

```diff
@@ -0,0 +1,270 @@
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
+
+
+# ----------------------------------------------------------------------
+# DataFrame: the main public API
+
+
+class DataFrame(ABC, Mapping):
+    """
+    An abstract data frame base class.
+
+    A "data frame" represents an ordered collection of named columns. A
+    column's "name" is permitted to be any hashable Python value, but strings
+    are common. Names are not required to be unique. Columns may be accessed by
+    name (when the name is unique) or by position.
+    """
+
+    def __dataframe__(self):
+        """
+        Idempotence of data frame protocol
+        """
+        return self
+
+    def __iter__(self):
+        # TBD: Decide what iterating should return
+        return iter(self.column_names)
+
+    def __len__(self):
+        return self.num_rows
+
+    @property
+    @abstractmethod
+    def num_columns(self):
+        """
+        Return the number of columns in the DataFrame
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def num_rows(self):
+        """
+        Return the number of rows in the DataFrame (if known)
+        """
+        pass
+
+    @abstractmethod
+    def iter_column_names(self) -> Iterable[Any]:
+        """
+        Return the column names as an iterable
+        """
+        pass
+
+    # TODO: Should this be a method or property?
+    @property
+    @abstractmethod
+    def column_names(self) -> Sequence[Any]:
+        """
+        Return the column names as a materialized sequence
+        """
+        pass
+
+    # TODO: Should this be a method or property?
+    @property
+    def row_names(self) -> Sequence[Any]:
+        """
+        Return the row names (if any) as a materialized sequence. It is not
+        necessary to implement this method
+        """
+        raise NotImplementedError("row_names")
+
+    def __getitem__(self, key: Hashable) -> Column:
```

Comment:
My preference would be to force column names to be `str`. I think the advantages in simplifying this are more than the limitation it imposes. Then `__getitem__` can unambiguously perform all reasonable operations, get one item (by name or position), get multiple, and indexing.

Reply:
I think we should be careful about going down the route of having a heavily-overloaded `__getitem__` in this interface. It may be better to be as explicit and unambiguous as possible. For example, I am not even sure that `df[start:end]` should be available

Reply:
This interface should perhaps not even have a `__getitem__` because it's really ambiguous and specific to implementations. It should be explicitly `get_columns` or `get_rows` to avoid ambiguity between implementations.

Reply:
Agree with you. I'd surely avoid too much overloading, but what is too much is difficult to say. May be some examples can help see what is our preference in practice:
```python
df.column_by_name('col2')
df.column_by_position(1)
df.columns_by_name('col2', 'col3')  # or receive a list instead of `*args`
...
df['col2']
df[1]
df['col2', 'col3']  # note that the param of `__getitem__` is a tuple, not a list like in df[['col2', 'col3']]
```
Not sure what's the best option, may be something in between, but I'm biased towards the second set. But surely avoiding any ambiguity or something too crazy. And I think requiring column names to be string, makes things easier. For example if a tuple is accepted (it is in your code, as it's hashable), things become more complex and easier to be ambiguous (see the last line of my example).

So, the exact API surely requires some discussion, but my main point was that I'd personally limit `key` to be a string. I understand that many people will prefer the flexibility over the simplicity, but that's my preference.

Reply:
> It should be explicitly `get_columns` or `get_rows` to avoid ambiguity between implementations.

I think this is a very good point. Do we want this PR to be both the API for end users and downstream projects accepting generic dataframes? From @devin-petersohn comment may be we should be discussing the API only for the latter (downstream projects), and leave the end user API to each project. If we do that, I'd avoid using any magic method, and prefix every method so there are no collisions with the public API of the implementations (e.g. `__dataframe_columns__`, `__dataframe_num_rows__`).

Reply:
My understanding of the discussion so far is that objects providing data frame-like data should have a single "export" method `__dataframe__` that returns an implementation of a "data frame interface" object. I do not think that e.g. `pandas.DataFrame` should be a subclass of this object. This PR is intending to define some of the APIs that this object should contain